magenta-lib/src/main/scala/magenta/tasks/ChangeSetTasks.scala

```diff
@@ -18,9 +19,9 @@ class CreateChangeSetTask(
                            val unresolvedParameters: CloudFormationParameters
 )(implicit val keyRing: KeyRing, artifactClient: S3Client) extends Task {
 
-  override def execute(reporter: DeployReporter, stopFlag: => Boolean) = if (!stopFlag) {
+  override def execute(reporter: DeployReporter, stopFlag: => Boolean, stsClient: StsClient) = if (!stopFlag) {
     CloudFormation.withCfnClient(keyRing, region) { cfnClient =>
-      S3.withS3client(keyRing, region) { s3Client =>
+      S3.withS3client(keyRing, region, stsClient = stsClient) { s3Client =>
         STS.withSTSclient(keyRing, region) { stsClient =>
```

Comment:
S3.withS3client should take a resources now (as it needs both a reporter to grab the deploy ID and also the stsClient). I would suggest that the artifactClient passed in on 19/20 here becomes a DeploymentResources. This is something we know at the time we create the task. Although to be fair it's not completely clear cut as that would provide two routes to the reporter.

Reply:
Needs more thought 🤔

Reply:
Hey @sihil , thanks for these comments - I only picked them up today. We ended up changing the signature of execute to take a `resources` instead of a `reporter` everywhere (see recent commits) - what do you think?

Reply:
I think that's a reasonable choice. I think it makes it a little easier to test the delpoyment types too.